client/modules/cm-lang-ftml/src/grammars/ftml.yaml

```diff
@@ -0,0 +1,999 @@
+    tag: tagName
+
+  BlockNameValue:
+    lookup: $var:blk_val
+    lookahead: /@namela/
+    emit: BlockName
+    tag: tagName
+
+  BlockNameValueMap:
+    lookup: $var:blk_valmap
+    lookahead: /@namela/
+    emit: BlockName
+    tag: tagName
+
+  BlockNameElements:
+    lookup: $var:blk_el
+    lookahead: /@namela/
+    emit: BlockName
+    tag: tagName
+
+  BlockNameMapElements:
+    lookup: $var:blk_map_el
+    lookahead: /@namela/
+    emit: BlockName
+    tag: tagName
+
+  BlockNameValueElements:
+    lookup: $var:blk_val_el
+    lookahead: /@namela/
+    emit: BlockName
+    tag: tagName
+
+  BlockNameAlign:
+    lookup: $var:blk_align
+    lookahead: /@namela/
+    tag: function(name)
+
+  BlockNameModule:
+    match: module
+    lookahead: /@namela/
+    tag: keyword
+
+  ModuleName:
+    lookup: $var:mods
+    lookahead: /@namela/
+    tag: className
+
+  ModuleNameUnknown:
+    match: /[^\\#*\s\]]+/
+    lookahead: /@namela/
+    tag: invalid
+
+  BlockNameUnknown:
+    match: /[^\\#*\s\]]+/
+    lookahead: /@namela/
+    tag: invalid
+
+  # nested blocks
+
+  BlockNodeInlineMath:
+    match: /(@BlockStart)(\$)([^\n]*?)(\$)(@BlockEnd)/
+    captures:
+      0: { is: BlockStart }
+      1: { type: BlockNodeInlineMathMark, tag: keyword }
+      2: { type: InlineMath, nest: wikimath }
+      3: { is: BlockNodeInlineMathMark }
+      4: { is: BlockEnd }
+
+  BlockNameMath:
+    match: math
+    lookahead: /@namela/
+    emit: BlockName
+    tag: special(tagName)
+
+  BlockNestedMath:
+    emit: BlockNested
+    nest: wikimath
+    fold: inside
+    begin:
+      type: BlockNestedMathStartNode
+      emit: BlockNode
+      indent: delimited(]])
+      skip: /\s+/
+      chain:
+        - BlockStart
+        - BlockNamePrefix?
+        - BlockNameMath
+        - BlockNameSuffix?
+        - BlockNodeArgument |* BlockLabel
+        - BlockEnd
+    end:
+      type: BlockNestedMathEndNode
+      emit: BlockNode
+      indent: delimited(]])
+      skip: /\s+/
+      chain:
+        - BlockStartClosing
+        - BlockNameMath
+        - BlockNameSuffix?
+        - BlockEnd
+
+  ModuleNameCSS:
+    match: css
+    lookahead: /@namela/
+    emit: BlockName
+    tag: className
+
+  BlockNestedModuleCSS:
+    emit: BlockNested
+    nest: CSS
+    fold: inside
+    begin:
+      type: BlockNestedModuleCSSStartNode
+      emit: BlockNode
+      indent: delimited(]])
+      skip: /\s+/
+      chain:
+        - BlockStart
+        - BlockNamePrefix?
+        - BlockNameModule
+        - BlockNameSuffix?
+        - ModuleNameCSS
+        - BlockLabel*
+        - BlockEnd
+    end:
+      type: BlockNestedModuleCSSEndNode
+      emit: BlockNode
+      indent: delimited(]])
+      skip: /\s+/
+      chain:
+        - BlockStartClosing
+        - BlockNameModule
+        - BlockNameSuffix?
+        - BlockEnd
+
+  BlockNameCSS:
+    match: css
+    lookahead: /@namela/
+    emit: BlockName
+    tag: special(tagName)
+
+  BlockNestedCSS:
```

Comment:
Why is CSS allowed to be nested?

Reply:
This is the rule for `[[css]]` blocks.

I made it a convention to name block containers that nest some other language `BlockNested`, mostly so that they're very obvious in the AST.

Reply:
Ah, okay.